Run model

lachmanfrantisek opened the pull request:
Introduce a `RunModel` that represents one pipeline.

Fixes #895 

Connects `JobTriggerModel` (and triggers like `PullRequestModel` via that model) with
build/test models like `SRPMBuildModel`, `CoprBuildModel`, `KojiBuildModel`, and `TFTTestRunModel`.

* One model of each build/test model can be connected.
* Each build/test model can be connected to multiple RunModels (e.g. on retrigger).
* Each RunModel has to be connected to exactly one `JobTriggerModel`.
* There can be multiple `RunModel`s for one `JobTriggerModel`.
  (e.g. For each push to PR, there will be a new `RunModel`, but the same `JobTriggerModel`.)


TODO:
- [x] Fix the tests.
- [x] Fix database tests
- [x] Improve/clean the database tests
- [x] Prepare migrations.
- [x] Test the migration on real data (db dump of prod/stage)
    - [x] stage
    - [x] prod
- [x] Add timestamp for `RunModel` and `TFTTestRunModel`.

TomasTomecek reviewed (commented):
reviewed very briefly, looks good!

jpopelka reviewed (commented):
nice!

lachmanfrantisek commented:
Migration on stage:

Test run matching:
```
Run models: tests_match_build=2304, more_tests=558, more_builds=8
Test runs: attached=2351, deleted=2687
```
Time:
```
real    0m40.516s
user    0m17.218s
sys     0m1.263s
```

TomasTomecek commented on alembic/versions/a5c06aa9ef30_add_runmodel.py:
```diff
@@ -0,0 +1,621 @@
+
+    session.commit()
+
+    # Remove direct connections:
+
+    op.drop_constraint(
+        "copr_builds_job_trigger_id_fkey", "copr_builds", type_="foreignkey"
+    )
+    op.drop_constraint(
+        "copr_builds_srpm_build_id_fkey1", "copr_builds", type_="foreignkey"
+    )
+    op.drop_column("copr_builds", "job_trigger_id")
+    op.drop_column("copr_builds", "srpm_build_id")
+    op.drop_constraint(
+        "koji_builds_srpm_build_id_fkey", "koji_builds", type_="foreignkey"
+    )
+    op.drop_constraint(
+        "koji_builds_job_trigger_id_fkey", "koji_builds", type_="foreignkey"
+    )
+    op.drop_column("koji_builds", "job_trigger_id")
+    op.drop_column("koji_builds", "srpm_build_id")
+    op.drop_constraint(
+        "srpm_builds_job_trigger_id_fkey", "srpm_builds", type_="foreignkey"
+    )
+    op.drop_column("srpm_builds", "job_trigger_id")
+    op.drop_constraint(
+        "tft_test_runs_job_trigger_id_fkey", "tft_test_runs", type_="foreignkey"
+    )
+    op.drop_column("tft_test_runs", "job_trigger_id")
```
is there a risk that the DB would become inconsistent if something goes wrong during runing these?

lachmanfrantisek replied:
TBH, I am not sure if there is some transaction going on during this. I would say it shouldn't if it works for the current state of DB. But I can test that on the real data right before the deployment.

lachmanfrantisek replied:
I can try raising an error during those steps locally and see what happens. (And maybe move the `commit` call to the very end of this function.)

lachmanfrantisek replied:
I can confirm now that in the case of `Exception` the changes are not applied.

TomasTomecek commented on packit_service/models.py:
```diff
@@ -540,14 +588,25 @@ def __repr__(self):
+    * Each RunModel has to be connected to exactly one JobTriggerModel.
+    * There can be multiple RunModels for one JobTriggerModel.
+      (e.g. For each push to PR, there will be new RunModel, but same JobTriggerModel.)
+    """
```
+1, very helpful

TomasTomecek commented on alembic/versions/a5c06aa9ef30_add_runmodel.py:
```diff
@@ -0,0 +1,621 @@
+    print(f"Run models with test run correctly set: {run_models_successful}")
+    print("================================")
+
+    # Check:
```
I know that you run these on the prod data which is giving us a great confidence that the migration is gonna be smooth, but! since exceptions are being raised below, what should we do if one of the exceptions get fired?

lachmanfrantisek replied:
The transaction is not committed in that state so we will not change anything and can revert deployment and fix the migration script.

TomasTomecek reviewed (commented):
~~Mind~~ Run-blowing

lachmanfrantisek commented:
And here is the output for stage:

```
================================
SRPM models without any build: 45
SRPM models without any run (RunModel created): 197
SRPM models removed because of no connection to any job trigger: 142
================================
All Copr builds: 10018
Copr builds deleted for no SRPM for trigger: 96
Copr builds deleted for no SRPM set: 627
Fixed SRPM matching to trigger model from Copr build: 0
================================
All Koji builds: 0
Koji builds deleted for no SRPM for trigger: 0
Koji builds deleted for no SRPM set: 0
Fixed SRPM matching to trigger model from Koji build: 0
================================
All Test runs: 5038
Attached correctly to build: 2390
All Run models: 9537
Run models with different number of tests and builds: 530
Run models with test run correctly set: 2340
================================
real    0m50.777s
user    0m21.050s
sys     0m1.670s
```

lachmanfrantisek commented:
I've made a small change how we get a build for a testing-farm run -- see the relevant commit: 23369fbc7019fc6f95faf8333fee9a0ee74d6516

TomasTomecek commented:
6+ weeks in works; you sir deserve an award for completing this :rocket: :tomato:
